### PR TITLE
[csv] use header option in batch loader

### PR DESCRIPTION
**Pull request by Pessimistress**
- Batch loader and sync loader now outputs consistent results (array if no header, object if header).
- Fix bugs in `ColumnarTableBatch`
- Tests

### CODE REVIEW

File: modules/experimental/src/categories/table/row-table-batch.js [hunk 1]
```diff
@@ -21,10 +28,23 @@ export default class RowTableBatch {
 
   getNormalizedBatch() {
     if (this.rows) {
-      const rows = this.rows.slice(0, this.length);
+      let rows = this.rows.slice(0, this.length);
+
+      if (this._headers) {
+        rows = rows.map(row => convertRowToObject(row, this._headers));
```
Comment (ibgreen) [inline modules/experimental/src/categories/table/row-table-batch.js]: This looks expensive! 

Feels like this should be done only on read...

File: modules/experimental/src/categories/table/row-table-batch.js [hunk 2]
```diff
@@ -4,6 +4,13 @@ export default class RowTableBatch {
     this.batchSize = batchSize;
     this.rows = null;
     this.length = 0;
+
+    if (!Array.isArray(schema)) {
```
Comment (ibgreen) [inline modules/experimental/src/categories/table/row-table-batch.js]: What happens if schema is an array? Add a comment?

File: modules/csv/src/csv-loader.js [hunk 3]
```diff
@@ -63,7 +63,8 @@ function parseCSVInBatches(asyncIterator, options) {
 
       // Check if we need to save a header row
       if (isFirstRow && !headerRow) {
-        if (isHeaderRow(row)) {
+        const {header = isHeaderRow(row)} = options;
```
Comment (ibgreen) [inline modules/csv/src/csv-loader.js]: Total nit: Just a bit too clever for my taste (i.e. optimize for readability not lines of code). I already felt the header handling was more complex than I would have preferred...

Would something like this work? 
```
if (options.header || isHeaderRow())
```

File: modules/experimental/src/categories/table/columnar-table-batch.js [hunk 4]
```diff
@@ -25,10 +25,18 @@ export default class ColumnarTableBatch {
   }
 
   getNormalizedBatch() {
-    const columns = this.columns;
+    this.pruneColumns();
+    const columns = Array.isArray(this.schema) ? this.columns : {};
+
+    if (!Array.isArray(this.schema)) {
```
Comment (ibgreen) [inline modules/experimental/src/categories/table/columnar-table-batch.js]: What happens if schema is an array?

Is `schema` required or can it be null?

Comment (Pessimistress) [reply modules/experimental/src/categories/table/columnar-table-batch.js]: `schema` is an array if there is no header row.

**Comment by Pessimistress** [comment]
> The primary concern is probably that we support conversion of CSV batches to arrow regardless of whether rows are encoded in array or object format

That is what the `ColumnarTableBatch` class does

**Pull request merged**